text/0028-generalized-entity-data-source.md

```diff
@@ -0,0 +1,524 @@
+`Entities`; they are stored as restricted expressions and evaluated to `Value`
+on every call to `is_authorized()`. (See
+[cedar#227](https://github.com/cedar-policy/cedar/issues/227).)
+This is inefficient.
+Today, `Entities::evaluate()` is implemented on `main` and provides a way to
+mitigate this issue, but it is opt-in and we'd rather have it be the default
+behavior.
+Notably, as of this writing, `Entities::evaluate()` is not released yet, so we
+can make changes in this area without breaking our users.
+
+**Challenge 4:**
+An additional motivation is that we would like to support partial evaluation
+better than the current `Entities::partial()` does, and the proposed
+`EntityDataSource` trait provides additional flexibility that should be helpful
+for this down the line.
+
+## Detailed design
+
+This RFC has four components:
+
+### Component 1: `is_authorized()` changes
+
+First, we generalize `is_authorized()` to accept any `EntityDataSource`, not
+just `Entities`:
+
+```rust
+impl Authorizer {
+    // status quo:
+    pub fn is_authorized(
+        &self,
+        q: &Request,
+        pset: &PolicySet,
+        entities: &Entities,
+    ) -> Response
+
+    // proposed change:
+    pub fn is_authorized(
+        &self,
+        q: &Request,
+        pset: &PolicySet,
+        entities: &impl EntityDataSource,
+    ) -> Response
+}
+```
+
+The trait `EntityDataSource` is defined as follows:
+```rust
+trait EntityDataSource {
+    /// Does an entity with the given UID exist or not
+    fn entity_exists(&self, uid: &EntityUid) -> Result<bool>;
+
+    /// Get just the indicated attribute of the specified entity, or `Ok(None)`
+    /// if the attribute does not exist.
+    ///
+    /// Implementations of this method should not assume that the entity exists.
+    /// If the entity does not exist, the implementation should return
+    /// `Err(UnknownEntity)` and not `Ok(None)`.
+    fn entity_attr(&self, uid: &EntityUid, attr: &str) -> Result<Option<Value>>;
+
+    /// Does the given attribute exist on the specified entity?
+    ///
+    /// Implementations of this method should not assume that the entity exists.
+    /// If the entity does not exist, the implementation should return
+    /// `Err(UnknownEntity)` and not `Ok(false)`.
+    fn entity_has_attr(&self, uid: &EntityUid, attr: &str) -> Result<bool> {
+        // default implementation in terms of `entity_attr()`.
+        // Note this computes the attribute value itself, and throws it away,
+        // which may be inefficient for some implementations.
+        Ok(self.entity_attr(uid, attr)?.is_some())
+    }
+
+    /// Is `child in parent`?
+    ///
+    /// Implementations of this method should not assume that either entity exists.
+    /// If either entity does not exist, the implementation should return
+    /// `Err(UnknownEntity)` and not `Ok(false)`.
+    fn entity_in(&self, child: &EntityUid, parent: &EntityUid) -> Result<bool>;
+}
+```
```

Comment:
A not very well thought out idea:

IIUC, if we have the expression `User::"alice".foo.bar.baz` and `User::"alice"` has record `foo` which has record `bar` which has bool `baz`, we require fetching all of `foo` to evaluate the expression. I think the motivation behind this RFC is that passing an entire Entity may be too expensive. Does that motivation apply to records?

If so, would it make sense to take multiple `attr`s? Something like:
```
    fn entity_attrs(&self, uid: &EntityUid, attrs: [&str]) -> Result<Option<Value>>
    fn entity_has_attrs(&self, uid: &EntityUid, attrs: [&str]) -> Result<bool>
```

If we did this, it could add more complications. We might want to allow the `EntityDataSource` to return the entire record at any point along the chain (e.g., in the above `foo` might be large but `bar` may be small and the policy could reference `User::"alice".foo.bar.baz1`, `User::"alice".foo.bar.baz2`...). Should the evaluator request `User::"alice".foo.bar` because it sees it will be used multiple times? Should the `EntityDataSource` return it because it knows `User::"alice".foo.bar` is small?